### ISSUE TITLE
Wrong config passed to a plugin when multiple tasks define different configs for a plugin and auto-start on creation

### ISSUE THREAD

**Issue by geauxvirtual** [intelsdi-x/snap#674]
OS Version: Mac OS X 10.10.5

Snap Version: v0.11.0-beta-8-g107e541 (HEAD)

Issue:
When creating multiple tasks that define different configs for the same plugin and the task auto starts (default behavior on task creation), the second task created will pass the plugin config information from the first task.

Note: I have only tested this a publisher plugin, but it's a good assumption this would occur for collect and processor plugins as well.

Here are the tasks that are loaded.

Task 1
![screen shot 2016-01-21 at 11 57 27 am](https://cloud.githubusercontent.com/assets/1395030/12492667/304ab9fc-c036-11e5-80cc-107e2849deab.png)

Task 2
![screen shot 2016-01-21 at 11 58 13 am](https://cloud.githubusercontent.com/assets/1395030/12492686/481b48b2-c036-11e5-8cba-d02f34b3f952.png)

The JSON output of the exported task via snapctl
```
{
    "id": "b68a5bad-4ebf-41e7-9f35-06ab59689146",
    "name": "rabRem",
    "deadline": "5s",
    "workflow": {
        "collect": {
            "metrics": {
                "/intel/psutil/load/load1": {
                    "version": 0
                },
                "/intel/psutil/load/load15": {
                    "version": 0
                },
                "/intel/psutil/load/load5": {
                    "version": 0
                },
                "/intel/psutil/vm/available": {
                    "version": 0
                },
                "/intel/psutil/vm/free": {
                    "version": 0
                },
                "/intel/psutil/vm/used": {
                    "version": 0
                }
            },
            "publish": [
                {
                    "plugin_name": "rabbitmq",
                    "plugin_version": 0,
                    "config": {
                        "address": "snappy:snappy@192.168.242.139:5672",
                        "exchange_name": "snap",
                        "exchange_type": "fanout",
                        "routing_key": "metrics"
                    }
                }
            ]
        }
    },
    "schedule": {
        "type": "simple",
        "interval": "10s"
    },
    "creation_timestamp": 1453405902,
    "last_run_timestamp": -1,
    "hit_count": 11,
    "task_state": "Stopped",
    "href": "http://localhost:8181/v1/tasks/b68a5bad-4ebf-41e7-9f35-06ab59689146",
    "Err": null
}
```

```
{
    "id": "938ce08f-c304-4af4-8b77-b7c1ab6ac530",
    "name": "rabLoc",
    "deadline": "5s",
    "workflow": {
        "collect": {
            "metrics": {
                "/intel/psutil/load/load1": {
                    "version": 0
                },
                "/intel/psutil/load/load15": {
                    "version": 0
                },
                "/intel/psutil/load/load5": {
                    "version": 0
                },
                "/intel/psutil/vm/available": {
                    "version": 0
                },
                "/intel/psutil/vm/free": {
                    "version": 0
                },
                "/intel/psutil/vm/used": {
                    "version": 0
                }
            },
            "publish": [
                {
                    "plugin_name": "rabbitmq",
                    "plugin_version": 0,
                    "config": {
                        "address": "localhost:5672",
                        "exchange_name": "snap",
                        "exchange_type": "fanout",
                        "routing_key": "metrics"
                    }
                }
            ]
        }
    },
    "schedule": {
        "type": "simple",
        "interval": "10s"
    },
    "creation_timestamp": 1453405910,
    "last_run_timestamp": -1,
    "hit_count": 12,
    "task_state": "Stopped",
    "href": "http://localhost:8181/v1/tasks/938ce08f-c304-4af4-8b77-b7c1ab6ac530",
    "Err": null
}
```

Screenshot of the snapd logs showing both tasks publishing to the same RabbitMQ instance.
![screen shot 2016-01-21 at 12 02 01 pm](https://cloud.githubusercontent.com/assets/1395030/12492805/e3747c2a-c036-11e5-84b7-b11fb1b22b23.png)

What I have found is this only occurs when tasks are auto started on creation. If you issue a --no-start on task creation and manually start the tasks, each task uses the correct config defined in the task. However, if you stop and start the auto started jobs, the tasks continue to persist with the wrong config.

**Comment by thomastaylor312** [comment]
I also saw this recently with some auto started tasks, but when I killed
and restarted the snap instance, it started publishing correctly.

**Comment by geauxvirtual** [comment]
Troubleshooting this some more, it seems that if the second task is created after the first task has run once, then the config for the second task is wrong. If the second task is created before the first task has run once, then the second task has the correct config.

**Comment by geauxvirtual** [comment]
After more debugging, I've tracked down the bug to these lines of code.

https://github.com/intelsdi-x/snap/blob/2f1e2e167508c7f7d93139abe6dea5d7e23595f0/control/control.go#L896-L903

What I noticed is that this issue always occurred after the first task had run once. This occurred for both when the task was auto-started and created with the --no-start flag. The second task would always have the same config as the first task.

When a task is created and dependencies are validated, the global config for a plugin will be merged against any config that is passed in the task manifest. In this scenario, there is no global config for the publisher plugin. When the first task is created, a key is created in the plugin cache with an empty data node. If the second task is created before the first is run, then it will merge an empty data node against it's plugin config. However, when either task is started and PublishMetrics is called, the code above merges the plugins config into the global config instead of the global config into the plugin config. Any task that is created after this occurs will merge the global config into it's plugin config on validation of dependencies. This should be a simple fix for both PublishMetrics and ProcessMetrics.
